src/data_type.hpp

```diff
@@ -0,0 +1,449 @@
+
+  CassValueType value_type() const { return value_type_; }
+
+  bool is_collection() const  {
+    return value_type_ == CASS_VALUE_TYPE_LIST ||
+        value_type_ == CASS_VALUE_TYPE_MAP ||
+        value_type_ == CASS_VALUE_TYPE_SET;
+  }
+
+  bool is_map() const  {
+    return value_type_ == CASS_VALUE_TYPE_MAP;
+  }
+
+  bool is_user_type() const {
+    return value_type_ == CASS_VALUE_TYPE_UDT;
+  }
+
+  bool is_tuple() const {
+    return value_type_ == CASS_VALUE_TYPE_TUPLE;
+  }
+
+  virtual bool equals(const SharedRefPtr<DataType>& data_type, bool exact) const {
+    if (exact) {
+      return value_type_ == data_type->value_type_;
+    } else {
+      switch (value_type_) {
+        case CASS_VALUE_TYPE_BIGINT:
+        case CASS_VALUE_TYPE_COUNTER:
+        case CASS_VALUE_TYPE_TIMESTAMP:
+          return is_int64_type(data_type->value_type_);
+
+        case CASS_VALUE_TYPE_ASCII:
+        case CASS_VALUE_TYPE_TEXT:
+        case CASS_VALUE_TYPE_VARCHAR:
+          return is_string_type(data_type->value_type_);
+
+        case CASS_VALUE_TYPE_BLOB:
+        case CASS_VALUE_TYPE_VARINT:
+          return is_bytes_type(data_type->value_type_);
+
+        case CASS_VALUE_TYPE_TIMEUUID:
+        case CASS_VALUE_TYPE_UUID:
+          return is_uuid_type(data_type->value_type_);
+
+        default:
+          return value_type_ == data_type->value_type_;
+      }
+    }
+  }
+
+private:
+  CassValueType value_type_;
+
+private:
+ DISALLOW_COPY_AND_ASSIGN(DataType);
+};
+
+typedef std::vector<SharedRefPtr<DataType> > DataTypeVec;
+
+class CollectionType : public DataType {
+public:
+  CollectionType(CassValueType collection_type)
+    : DataType(collection_type) { }
+
+  CollectionType(CassValueType collection_type, const DataTypeVec& types)
+    : DataType(collection_type)
+    , types_(types) { }
+
+  DataTypeVec& types() { return types_; }
+  const DataTypeVec& types() const { return types_; }
+
+  virtual bool equals(const SharedRefPtr<DataType>& data_type, bool exact) const {
+    assert(value_type() == CASS_VALUE_TYPE_LIST ||
+           value_type() == CASS_VALUE_TYPE_SET ||
+           value_type() == CASS_VALUE_TYPE_MAP ||
+           value_type() == CASS_VALUE_TYPE_TUPLE);
+
+    if (value_type() != data_type->value_type()) {
+      return false;
+    }
+
+    const SharedRefPtr<CollectionType>& collection_type(data_type);
+
+    if (exact) {
+      if (types_.size() != collection_type->types_.size()) {
+        return false;
+      }
+
+      for (size_t i = 0; i < types_.size(); ++i) {
+        if(!types_[i]->equals(collection_type->types_[i], true)) {
+          return false;
+        }
+      }
+    } else if(!types_.empty() && !collection_type->types_.empty()) { // An empty collection doesn't have subtypes
+      if (types_.size() != collection_type->types_.size()) {
+        return false;
+      }
+      for (size_t i = 0; i < types_.size(); ++i) {
+        if(!types_[i]->equals(collection_type->types_[i], false)) {
```

Comment:
I think if-exact-else could be simplified removing the conditional and passing `exact` instead of the bool literal.